### PR TITLE
fix: Reduce memory usage of gradient norm computation

### PR DESCRIPTION
**Pull request by jseppanen**
# What does this PR do ?

Reduce training memory usage by avoiding to materialize temporary float32 copies of all gradients

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved gradient norm calculation with dtype-consistent accumulation for all norm orders, including infinity norm, yielding more accurate and stable results across dtypes on CUDA.
  * Ensured consistent reductions across data/model parallel groups and removed unintended per-parameter casting side effects.
  * No changes to public APIs or clipping behavior.

* **Tests**
  * Added CUDA-only numeric precision tests covering multiple gradient/norm dtypes and norm orders (1, 2, ∞) to validate accuracy.
  * Isolated tests from distributed reductions to ensure deterministic verification.

### CODE REVIEW

File: nemo_rl/models/dtensor/parallelize.py [hunk 1]
```diff
@@ -807,19 +802,19 @@ def get_grad_norm(
         total_norm = float(total_norm_cuda[0].item())
 
     else:
-        total_norm = torch.tensor(0.0, dtype=torch.float32, device="cuda")
+        total_norm_cuda = torch.tensor(0.0, dtype=dtype, device="cuda")
         for grad in grads_for_norm:
-            grad_norm = torch.norm(grad, norm_type)
-            total_norm += torch.pow(grad_norm, norm_type)
+            grad_norm = torch.linalg.vector_norm(grad, ord=norm_type, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (jseppanen) [reply nemo_rl/models/dtensor/parallelize.py:808]: torch version is pinned to 2.7 and torch.norm is deprecated

File: nemo_rl/models/dtensor/parallelize.py [hunk 2]
```diff
@@ -807,19 +802,19 @@ def get_grad_norm(
...
-            total_norm, op=torch.distributed.ReduceOp.SUM, group=dp_cp_group
+            total_norm_cuda, op=torch.distributed.ReduceOp.SUM, group=dp_cp_group
         )
 
         torch.distributed.all_reduce(
-            total_norm, op=torch.distributed.ReduceOp.SUM, group=tp_group
+            total_norm_cuda, op=torch.distributed.ReduceOp.SUM, group=tp_group
         )
-        total_norm = total_norm.item() ** (1.0 / norm_type)  # type: ignore
+        total_norm = float(total_norm_cuda.item() ** (1.0 / norm_type))
```
[the remark this answers is not in the document]
Comment (jseppanen) [reply nemo_rl/models/dtensor/parallelize.py:820]: grads are always on CUDA

**Comment by joyang-nv** [comment]
I will finish the review tomorrow.

**Comment by joyang-nv** [comment]
@jseppanen can you rebase (update branch)? Since you have another merged change on the same function, I'd like to review based on updated version.

**Comment by jseppanen** [comment]
@terrykong @joyang-nv rebased and signed off

**Review by joyang-nv — APPROVED** [review]
Thanks for changing this. :)
For `torch.inf` norm, this completely remove type conversion and save memory and computation.
